pkg/storage/stores/composite_store.go

```diff
@@ -206,6 +208,83 @@ func (c CompositeStore) Volume(ctx context.Context, userID string, from, through
 	return res, err
 }
 
+func (c CompositeStore) GetShards(
+	ctx context.Context,
+	userID string,
+	from, through model.Time,
+	targetBytesPerShard uint64,
+	predicate chunk.Predicate,
+) ([]logproto.Shard, error) {
+	// TODO(owen-d): improve. Since shards aren't easily merge-able,
+	// we choose the store which returned the highest shard count.
+	// This is only used when a query crosses a schema boundary
+	var groups [][]logproto.Shard
+	err := c.forStores(ctx, from, through, func(innerCtx context.Context, from, through model.Time, store Store) error {
+		shards, err := store.GetShards(innerCtx, userID, from, through, targetBytesPerShard, predicate)
+		if err != nil {
+			return err
+		}
+		groups = append(groups, shards)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case len(groups) == 1:
+		return groups[0], nil
+	case len(groups) == 0:
+		return nil, nil
+	default:
+		sort.Slice(groups, func(i, j int) bool {
```

Comment:
Since you are looking for the biggest group, you don't need to sort this slice. Instead, you can use [slices.MinFunc][1] O(n).

[1]: https://pkg.go.dev/slices#MinFunc